models/src/main/scala/coop/rchain/models/HashM.scala

```diff
@@ -88,14 +86,15 @@ object HashM extends HashMDerivation {
 
   }
 
-  implicit val ParHash: HashM[Par] = gen[Par]
-  implicit val ExprHash            = gen[Expr]
-  implicit val VarHash             = gen[Var]
-  implicit val SendHash            = gen[Send]
-  implicit val ReceiveHash         = gen[Receive]
-  implicit val ReceiveBindHash     = gen[ReceiveBind]
-  implicit val NewHash             = gen[New]
-  implicit val MatchHash           = gen[Match]
+  implicit val Env: HashM[Map[String, Par]] = opaqueHash
```

Comment:
Please see this discussion: https://github.com/rchain/rchain/pull/2763#discussion_r335401269
If possible we should stick to `HashM[NormalizerEnv]`.

Reply:
I read that - but I don't understand still. Please explain - `NornalizerEnv` was just a type alias before

Reply:
It's just a little bit safer in terms of stack-safety because if someone one day creates a `map: Map[String, Par]` with some really big `Par` inside, Scala will pick up the `opaqueHash` instance and will blow up the stack.
With `HashM[NormalizerEnv]` that someone won't be able to do that. At least not without writing the non stack-safe `HashM` instance himself.

Reply:
Agreed